docs/scripts/make-all.rst

```diff
@@ -0,0 +1,181 @@
+  Supplies the value of the ``HOST_DEPEND_DIR`` listfile variable.
+  Defaults to the value of the ``HOST_INSTALL`` environment variable.
+
+``--ovs=OVS``, ``-O OVS``
+  Directory in which Open vSwitch will be (or is) installed.
+
+  May be omitted if the ``--no-ovs`` option is specified.
+  Specifies the value of ``CMAKE_INSTALL_PREFIX`` for the OVS build.
+  Supplies the value of the ``OVS_INSTALL_DIR`` listfile variable for
+  the networking recipe (P4 Control Plane) build.
+  May be the same as the ``prefix`` directory, in which case OVS and
+  P4 Control Plane will be installed to the same directory tree.
+  Defaults to the value of the ``OVS_INSTALL`` environment variable.
+
+``--prefix=PREFIX``, ``-P PREFIX``
+  Directory in which P4 Control Plane will be installed.
+
+  The directory will be created if it does not exist.
+  May be the same as the ``--ovs`` option, in which case OVS and
+  P4 Control Plane will be installed to the same directory tree.
+  Specifies the value of the ``CMAKE_INSTALL_PREFIX`` variable when
+  building P4 Control Plane.
+
+``--sde=SDE``, ``-S SDE``
+  Directory in which the SDK for the Intel&reg; E2100 IPU is installed.
+
+  Supplies the value of the ``SDE_INSTALL_DIR`` listfile variable.
+  Defaults to the value of the ``SDE_INSTALL`` environment variable.
+
+``--toolchain=FILE``, ``-T FILE``
+  Path to the CMake toolchain file.
+
+  Must be specified when cross-compiling.
+  Specifies the value of the ``CMAKE_TOOLCHAIN_FILE`` variable.
+  Defaults to the value of the ``CMAKE_TOOLCHAIN_FILE`` environment variable.
+
+Options
+-------
+
+``--coverage``
+  Instrument build to measure unit test code coverage.
+  Sets the ``TEST_COVERAGE`` listfile option to FALSE.
+
+``--cxx=STD``
+  C++ standard to be used by the compiler (11, 14, 17).
+  Specifies the value of the ``CXX_STANDARD`` listfile variable.
+
+``--jobs=NJOBS``, ``-j NJOBS``
+  Number of build threads.
+  Specifies the value of the ``-j`` CMake option.
+  Defaults to 8 threads.
+
+``--no-build``
+  Configures CMake but does not build the dependencies.
+
+``--no-krnlmon``
+  Excludes the Kernel Monitor from the build.
+  Sets the ``WITH_KRNLMON`` listfile option to FALSE.
+
+``--no-ovs``
+  Excludes OVS from the build.
+  Sets the ``WITH_OVSP4RT`` listfile option to FALSE.
+
+``--target=TARGET``
+  Specifies the target (``dpdk``, ``es2k``, or ``tofino``) for which
+  P4 Control Plane will be built.
+  The target name is case-insensitive.
+  Sets the ``TDI_TARGET`` listfile variable.
+  Defaults to ``DPDK`` if unspecified.
+
+Configurations
+--------------
+
+``--debug``
+  Build with ``-DCMAKE_BUILD_TYPE=Debug``.
+  The compiler settings will default to ``-g``.
+
+``--minsize``  
+  Build with ``-DCMAKE_BUILD_TYPE=MinSizeRel``.
+  The compiler settings will default to ``-Os -DNDEBUG``.
+
+``--reldeb``
+  Build with ``-DCMAKE_BUILD_TYPE=RelWithDebInfo``.
+  The compiler settings will default to ``-O2 -g -DNDEBUG``.
+
+``--release``
```

Comment:
What will be the default value if nothing is specified? I am trying to understand if we need to modify script where CI images are built, or default config is the one that is without symbols and has less memory footprint?

Reply:
If nothing is specified, the helper scripts should not specify CMAKE_BUILD_TYPE when they configure the build, letting the cmake listfile use its own default value.

The default value has been RelWithDebInfo for a long time.

I've been thinking lately that we ought to change the default to Release, which will boost the optimization level from `-O2` to `-O3` in addition to omitting the symbols. In a production system, the default build type _ought_ to be Release. Among other things, the Intel security guide dictates that you use some additional compiler settings to make the resulting images less vulnerable.

(We're not quite there yet -- it turns out that one or more of the recommended settings cause in link-time errors if the options differ between our code and the libraries we link to. It's going to take trial and error to figure out which of the options are causing the problem and what we need to do differently when we build the libraries -- assuming that we're the ones who build them.)

Making this change would affect anybody who expect to be able to run `gdb`. They'd have to start specifying `--reldeb` in order to maintain parity. We would therefore want to (a) announce the planned change in advance, and then (b) advertise the change prominently when we make it.

Another way to reduce the file size is to run `strip` on the image after building it with `RelWithDebInfo`. There is also a way to have the toolchain write the debug symbols to a separate file, which you can load or reference in your debugger. Neither of these would give you `-O3`.

Reply:
Good point, though. The document should say what the default behavior is.